llvm-passes/ObfuscateZero/ObfuscateZero.cpp

```diff
@@ -0,0 +1,186 @@
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/IPO/PassManagerBuilder.h"
+
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IRBuilder.h"
+
+#ifndef NDEBUG
+#include "llvm/IR/Verifier.h"
+#include "llvm/Support/Debug.h"
+#endif
+
+#include <vector>
+#include <random>
+
+using namespace llvm;
+
+namespace {
+class ObfuscateZero : public BasicBlockPass {
+  std::vector<Value *> IntegerVect;
+  std::default_random_engine Generator;
+
+public:
+  static char ID;
+
+  ObfuscateZero() : BasicBlockPass(ID) {}
+
+  virtual bool runOnBasicBlock(BasicBlock &BB) override {
+    IntegerVect.clear();
+    bool modified = false;
+
+    // Not iterating from the beginning to avoid obfuscation of Phi instructions
+    // parameters
+    for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),
+                                       end = BB.end();
+         I != end; ++I) {
+      Instruction &Inst = *I;
+      if (isValidCandidateInstruction(Inst)) {
+        for (size_t i{0}; i < Inst.getNumOperands(); ++i) {
+          if (Constant *C = isValidCandidateOperand(Inst.getOperand(i))) {
+            if (Value *New_val = replaceZero(Inst, C)) {
+              Inst.setOperand(i, New_val);
+              modified = true;
+            } else {
+              //dbgs() << "ObfuscateZero: could not rand pick a variable for replacement\n";
+            }
+          }
+        }
+      }
+      registerInteger(Inst);
+    }
+
+#ifndef NDEBUG
+    verifyFunction(*BB.getParent());
+#endif
+    return modified;
+  }
+
+private:
+  bool isValidCandidateInstruction(Instruction &Inst) {
+    if (isa<GetElementPtrInst>(&Inst)) {
+      // dbgs() << "Ignoring GEP\n";
+      return false;
+    } else if (isa<SwitchInst>(&Inst)) {
+      // dbgs() << "Ignoring Switch\n";
+      return false;
+    } else if (isa<CallInst>(&Inst)) {
+      // dbgs() << "Ignoring Calls\n";
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+  Constant *isValidCandidateOperand(Value *V) {
+    if (Constant *C = dyn_cast<Constant>(V)) {
+      // Checking constant eligibility
+      if (isa<PointerType>(C->getType())) {
+        // dbgs() << "Ignoring NULL pointers\n";
+        return nullptr;
+      } else if (C->getType()->isFloatingPointTy()) {
+        // dbgs() << "Ignoring Floats 0\n";
+        return nullptr;
+      } else if (C->isNullValue()) {
+        return C;
+      } else {
+        return nullptr;
+      }
+    } else {
+      return nullptr;
+    }
+  }
+
+  void registerInteger(Value &V) {
+    if (V.getType()->isIntegerTy())
+      IntegerVect.emplace_back(&V);
+  }
+
+  Value *replaceZero(Instruction &Inst, Value *VReplace) {
+    // Replacing 0 by:
+    // prime1 * ((x | any1)**2) != prime2 * ((y | any2)**2)
+    // with prime1 != prime2 and any1 != 0 and any2 != 0
+    using prime_type = uint32_t;
+    // FIXME : replace by dynamic generation
+    const prime_type p1 = 431, p2 = 277;
```

Comment:
Given the small size of the considered primes, you could get a random value, set it to be odd (``| 1``) then test for primality, add 2, test again etc. Not really random but fair enough for our needs.